ogr/parsing.py

```diff
@@ -71,6 +71,98 @@ def __str__(self) -> str:
         repo_url_str += ")"
         return repo_url_str
 
+    def __repr__(self) -> str:
+        return str(self)
+
+    @staticmethod
+    def _prepare_url(potential_url: str) -> Optional[ParseResult]:
+        # Remove trailing '/' from 1-7 to ensure parsing works as expected.
+        if potential_url[-1] == "/":
+            potential_url = potential_url.rstrip("/")
+
+        # transform 4-6 to a URL-like string, so that we can handle it together with 1-3
+        if "@" in potential_url:
+            split = potential_url.split("@")
+            if len(split) == 2:
+                potential_url = "http://" + split[1]
+            else:
+                # more @s ?
+                return None
+
+        # make it parsable by urlparse if it doesn't contain scheme
+        if not potential_url.startswith(
+            ("http://", "https://", "git://", "git+https://")
+        ):
+            potential_url = "http://" + potential_url
+
+        return urlparse(potential_url)
+
+    def _set_hostname_and_scheme(self, parsed_url: ParseResult):
+        self.hostname = parsed_url.hostname
+        self.scheme = parsed_url.scheme
+
+    def _parse_username(self, parsed: ParseResult) -> None:
+        if ":" not in parsed.netloc:
+            return
+
+        # e.g. domain.com:foo or domain.com:1234, where foo is username, but 1234 is port number
+        split = parsed.netloc.split(":")
+        if split[1] and not split[1].isnumeric():
+            self.username = split[1]
+
+    @staticmethod
+    def _prepare_path(parsed: ParseResult) -> Tuple[str, List[str]]:
+        # path starts with '/', strip it away
+        path = parsed.path.lstrip("/")
+
+        # strip trailing '.git'
+        if path.endswith(".git"):
+            path = path[: -len(".git")]
+
+        return path, path.split("/")
+
+    def _parsed_path(self, path: str, splits: List[str]) -> Optional["RepoUrl"]:
+        if len(splits) == 1:
+            self.namespace = self.username
+            self.repo = path
+
+            return self
+
+        if self.username or len(splits) < 2:
+            # all other cases (invalid)
+            return None
+
+        # path contains username/reponame
+        # or some/namespace/reponame
+        # or fork/username/some/namespace/reponame
+        self.is_fork = "fork" == splits[0] and len(splits) >= 3  # pagure fork
+        if self.is_fork:
+            # fork/username/namespace/repo format
+            self.username = splits[1]
+            namespace_parts = splits[2:-1]
+        else:
+            namespace_parts = splits[:-1]
+
+        self.namespace = "/".join(namespace_parts)
+        self.repo = splits[-1]
+
+        return self
+
+    @classmethod
+    def _parse(cls, potential_url: str) -> Optional["RepoUrl"]:
```

Comment:
It does not need to be private.

```suggestion
    def parse(cls, potential_url: str) -> Optional["RepoUrl"]:
```